File: packages/firebase_storage/firebase_storage/ios/firebase_storage/Sources/firebase_storage/DataStreamChannelStreamHandler.swift
// Copyright 2025 The Chromium Authors.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif

final class DataStreamChannelStreamHandler: NSObject, FlutterStreamHandler {
  private let downloadUrl: String
  private let maxSize: Int64
  private var task: URLSessionDataTask?
  private var isCancelled = false

  init(downloadUrl: String, maxSize: Int64) {
    self.downloadUrl = downloadUrl
    self.maxSize = maxSize
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    isCancelled = false
    guard let url = URL(string: downloadUrl) else {
      events([
        "error": [
          "code": "invalid-argument",
          "message": "Invalid download URL",
        ],
      ])
      return nil
    }

    let session = URLSession.shared
    task = session.dataTask(with: url) { [weak self] data, response, error in
      guard let self = self, !self.isCancelled else { return }

      if let error = error {
        events([
          "error": [
            "code": "unknown",
            "message": error.localizedDescription,
          ],
        ])
        return
      }

      guard let httpResponse = response as? HTTPURLResponse else {
        events([
          "error": [
            "code": "unknown",
            "message": "Invalid response type",
          ],
        ])
        return
      }

      if httpResponse.statusCode != 200 {
        events([
          "error": [
            "code": "unknown",
            "message": "Failed to download file: HTTP \(httpResponse.statusCode)",
          ],
        ])
        return
      }

      guard let data = data else {
        events([
          "error": [
            "code": "unknown",
            "message": "No data received",
          ],
        ])
        return
      }

      // Stream data in chunks
      let chunkSize = 8192  // 8KB chunks
      var totalBytesRead: Int64 = 0
      var offset = 0

      while offset < data.count && !self.isCancelled {
        let remainingBytes = data.count - offset
        let currentChunkSize = min(chunkSize, remainingBytes)
        let chunk = data.subdata(in: offset..<(offset + currentChunkSize))

        totalBytesRead += Int64(currentChunkSize)
        if self.maxSize > 0 && totalBytesRead > self.maxSize {
          events([
            "error": [
              "code": "download-size-exceeded",
              "message":
                "Downloaded data exceeds maximum allowed size of \(self.maxSize) bytes",
            ],
          ])
          return
        }

        events(["data": FlutterStandardTypedData(bytes: chunk)])
        offset += currentChunkSize
      }

      if !self.isCancelled {
        events(["complete": true])
      }
    }

    task?.resume()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    isCancelled = true
    task?.cancel()
    task = nil
    return nil
  }
}

File: packages/firebase_storage/firebase_storage/android/src/main/kotlin/io/flutter/plugins/firebase/storage/DataStreamChannelStreamHandler.kt
/*
 * Copyright 2025, the Chromium project authors.
 * Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
 */
package io.flutter.plugins.firebase.storage

import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

internal class DataStreamChannelStreamHandler(
  private val downloadUrl: String,
  private val maxSize: Long
) : StreamHandler {
  private val isCancelled = AtomicBoolean(false)
  private var scope: CoroutineScope? = null

  override fun onListen(arguments: Any?, events: EventSink) {
    isCancelled.set(false)
    scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    
    scope?.launch {
      try {
        val url = URL(downloadUrl)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connect()
        
        if (connection.responseCode != HttpURLConnection.HTTP_OK) {
          val errorMap = mapOf(
            "code" to "unknown",
            "message" to "Failed to download file: HTTP ${connection.responseCode}"
          )
          events.success(mapOf("error" to errorMap))
          return@launch
        }
        
        val inputStream: InputStream = connection.inputStream
        val buffer = ByteArray(8192) // 8KB chunks
        var totalBytesRead = 0L
        
        try {
          while (!isCancelled.get()) {
            val bytesRead = inputStream.read(buffer)
            if (bytesRead == -1) {
              // End of stream
              events.success(mapOf("complete" to true))
              break
            }
            
            totalBytesRead += bytesRead
            if (maxSize > 0 && totalBytesRead > maxSize) {
              val errorMap = mapOf(
                "code" to "download-size-exceeded",
                "message" to "Downloaded data exceeds maximum allowed size of $maxSize bytes"
              )
              events.success(mapOf("error" to errorMap))
              break
            }
            
            val chunk = buffer.copyOf(bytesRead)
            events.success(mapOf("data" to chunk))
          }
        } finally {
          inputStream.close()
          connection.disconnect()
        }
      } catch (e: Exception) {
        if (!isCancelled.get()) {
          val errorMap = FlutterFirebaseStoragePlugin.getExceptionDetails(e)
          events.success(mapOf("error" to errorMap))
        }
      }
    }
  }

  override fun onCancel(arguments: Any?) {
    isCancelled.set(true)
    scope?.cancel()
    scope = null
  }
}

